e2e_playwright/shared/app_utils.py

```diff
@@ -503,3 +503,127 @@ def check_top_level_class(app: Page, test_id: str) -> None:
         The test id of the element to check.
     """
     expect(app.get_by_test_id(test_id).first).to_have_class(re.compile(test_id))
+
+
+def register_connection_status_observer(page_or_frame: Page | Frame | None) -> None:
+    if page_or_frame is None:
+        return None
+
+    return page_or_frame.evaluate("""async () => {
+        window.streamlitPlaywrightDebugConnectionStatuses = [];
+        const callback = (mutationList, observer) => {
+            if (!mutationList || mutationList.length === 0) {
+                return
+            }
+            const target = mutationList[0].target
+            if (!target) {
+                return
+            }
+            let state = target
+                            .getAttribute('data-test-connection-state')
+                            .toUpperCase();
+            window.streamlitPlaywrightDebugConnectionStatuses.push(state);
+        }
+        const observer = new MutationObserver(callback);
+        // Observe app status for changes
+        const targetNode = document.querySelector('[data-testid=stApp]')
+        if (!targetNode) {
+            console.log("stApp not found")
+            return
+        }
+        const config = {
+            childList: false,
+            subtree: false,
+            attributeFilter: ['data-test-connection-state']
+        };
+        observer.observe(targetNode, config);
+    }""")
+
+
+def get_observed_connection_statuses(page_or_frame: Page | Frame | None) -> list[str]:
+    if page_or_frame is None:
+        return []
+
+    return page_or_frame.evaluate(
+        "() => window.streamlitPlaywrightDebugConnectionStatuses"
+    )
+
+
+def get_connection_status(
```

Comment:
As commented below: Maybe this can just be an `expect_connection_status` method that doesn't return anything. It would make the usage slightly simpler.